File: modules/signal-protocol/android/src/main/java/expo/modules/signalprotocol/EncryptedSession.kt
package expo.modules.signalprotocol

import android.content.Context
import expo.modules.signalprotocol.stores.*
import org.signal.libsignal.protocol.SessionBuilder
import org.signal.libsignal.protocol.SessionCipher
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.KyberPreKeyStore
import org.signal.libsignal.protocol.state.PreKeyBundle
import org.signal.libsignal.protocol.state.PreKeyStore
import org.signal.libsignal.protocol.state.SessionStore
import org.signal.libsignal.protocol.state.SignedPreKeyStore

/**
 * EncryptedSession - Manages encrypted communication with a remote user.
 *
 * Uses shared stores for pre-keys (H6/H7 fix) so that all sessions see
 * new pre-keys immediately after replenishment/rotation.
 *
 * @param localUser The local user's identity and keys
 * @param remoteAddress The remote user's Signal protocol address
 * @param remoteUser The remote user's public keys (null when resuming existing session)
 * @param storageId Stable identifier for per-session storage (should be remoteUserId)
 * @param sharedPreKeyStore Shared pre-key store from plugin
 * @param sharedSignedPreKeyStore Shared signed pre-key store from plugin
 * @param sharedKyberPreKeyStore Shared Kyber pre-key store from plugin
 * @param androidContext Android context for storage access
 */
class EncryptedSession(
    private val localUser: LocalUser,
    val remoteAddress: SignalProtocolAddress,
    private val remoteUser: RemoteUser?,
    private val storageId: String,
    private val sharedPreKeyStore: PreKeyStore,
    private val sharedSignedPreKeyStore: SignedPreKeyStore,
    private val sharedKyberPreKeyStore: KyberPreKeyStore,
    androidContext: Context
) {
    // Per-session stores (track state with this specific remote user)
    internal val identityStore: PersistentIdentityKeyStore
    private val sessionStore: SessionStore

    // Shared stores (all sessions share these - our pre-keys)
    private val preKeyStore: PreKeyStore = sharedPreKeyStore
    private val signedPreKeyStore: SignedPreKeyStore = sharedSignedPreKeyStore
    private val kyberPreKeyStore: KyberPreKeyStore = sharedKyberPreKeyStore

    init {
        // Use storageId (remoteUserId) as keyPrefix for per-session stores
        val keyPrefix = storageId
        identityStore = PersistentIdentityKeyStore(localUser.identityKey, localUser.registrationId, keyPrefix, androidContext)
        sessionStore = PersistentSessionStore(keyPrefix, androidContext)

        // Process remote user's preKey bundle if this is a new session
        if (remoteUser != null) {
            processRemoteBundle(remoteUser)
        }
    }

    private fun processRemoteBundle(remoteUser: RemoteUser) {
        val preKeyBundle = PreKeyBundle(
            remoteUser.registrationId,
            remoteAddress.deviceId,
            remoteUser.preKeyId,
            remoteUser.preKeyPublicKey,
            remoteUser.signedPreKeyId,
            remoteUser.signedPreKeyPublicKey,
            remoteUser.signedPreKeySignature,
            remoteUser.identityKeyPairPublicKey,
            remoteUser.kyberPreKeyId,
            remoteUser.kyberPreKeyPublicKey,
            remoteUser.kyberPreKeySignature
        )

        val sessionBuilder = SessionBuilder(
            sessionStore, preKeyStore, signedPreKeyStore, identityStore, remoteAddress
        )
        sessionBuilder.process(preKeyBundle)
    }

    fun encrypt(message: String): String {
        val messageBytes = message.toByteArray(Charsets.UTF_8)
        val sessionCipher = SessionCipher(sessionStore, preKeyStore, signedPreKeyStore, kyberPreKeyStore, identityStore, remoteAddress)
        val cipherTextMessage = sessionCipher.encrypt(messageBytes)
        return android.util.Base64.encodeToString(cipherTextMessage.serialize(), android.util.Base64.NO_WRAP)
    }

    fun decrypt(encryptedMessage: String): String? {
        val messageData = android.util.Base64.decode(encryptedMessage, android.util.Base64.NO_WRAP)
        val sessionCipher = SessionCipher(sessionStore, preKeyStore, signedPreKeyStore, kyberPreKeyStore, identityStore, remoteAddress)
        return try {
            val signalMsg = SignalMessage(messageData)
            val decrypted = sessionCipher.decrypt(signalMsg)
            String(decrypted, Charsets.UTF_8)
        } catch (e: Exception) {
            val preKeySignalMsg = PreKeySignalMessage(messageData)
            try {
                val decrypted = sessionCipher.decrypt(preKeySignalMsg)
                String(decrypted, Charsets.UTF_8)
            } catch (preKeyError: Exception) {
                // PreKey may have been consumed by a previous message in this batch.
                // The first PreKeySignalMessage establishes the session via decrypt(PreKeySignalMessage),
                // but subsequent messages from the same sender (before they receive our reply)
                // still arrive as PreKeySignalMessages. Since the session is already established,
                // we can extract the inner SignalMessage and decrypt it with the existing session.
                val innerSignalMsg = preKeySignalMsg.getWhisperMessage()
                val decrypted = sessionCipher.decrypt(innerSignalMsg)
                String(decrypted, Charsets.UTF_8)
            }
        }
    }
}

File: modules/signal-protocol/ios/EncryptedSession.swift
import Foundation
import LibSignalClient

class EncryptedSession {

    final let localUser: LocalUser
    final var remoteUser: RemoteUser?
    final let remoteUserProtocolAddress: ProtocolAddress
    final let storageDirectory: URL

    // Per-session stores (track state with this specific remote user)
    internal let identityStore: PersistentIdentityKeyStore
    private let sessionStore: SessionStore

    // Shared stores (all sessions share these - our pre-keys)
    private let preKeyStore: PreKeyStore
    private let signedPreKeyStore: SignedPreKeyStore
    private let kyberPreKeyStore: KyberPreKeyStore

    private let context = NullContext()

    /// Initialize session with shared stores
    /// - Parameters:
    ///   - localUser: The local user
    ///   - remoteUser: The remote user (nil for resumed sessions)
    ///   - protocolAddress: Remote user's protocol address
    ///   - storageId: Stable identifier for per-session storage (should be remoteUserId)
    ///   - sharedPreKeyStore: Shared pre-key store from plugin
    ///   - sharedSignedPreKeyStore: Shared signed pre-key store from plugin
    ///   - sharedKyberPreKeyStore: Shared Kyber pre-key store from plugin
    init(
        localUser: LocalUser,
        remoteUser: RemoteUser?,
        protocolAddress: ProtocolAddress,
        storageId: String,
        sharedPreKeyStore: PreKeyStore,
        sharedSignedPreKeyStore: SignedPreKeyStore,
        sharedKyberPreKeyStore: KyberPreKeyStore
    ) throws {
        self.localUser = localUser
        self.remoteUser = remoteUser
        self.remoteUserProtocolAddress = protocolAddress
        self.storageDirectory = EncryptedSession.getStorageDirectoryFor(remoteUserId: storageId)

        // Shared stores (pre-keys are ours, shared across all sessions)
        self.preKeyStore = sharedPreKeyStore
        self.signedPreKeyStore = sharedSignedPreKeyStore
        self.kyberPreKeyStore = sharedKyberPreKeyStore

        // Per-session stores (track state with this specific remote user)
        self.sessionStore = PersistentSessionStore(directoryURL: storageDirectory)
        self.identityStore = PersistentIdentityKeyStore(
            identityKeyPair: localUser.identityKey,
            registrationId: localUser.registrationId,
            directoryURL: storageDirectory
        )

        // Process remote user's preKey bundle if this is a new session
        if let remoteUser = remoteUser {
            try processRemoteBundle(remoteUser)
        }
    }

    private static func getStorageDirectoryFor(remoteUserId: String) -> URL {
        // Return a placeholder URL with remoteUserId as last path component
        // The stores now use Keychain and only need the last path component as prefix
        return URL(fileURLWithPath: remoteUserId)
    }

    public func encrypt(message: String) throws -> String {
        let messageBytes: [UInt8] = [UInt8](Data(message.utf8))
        let cipherTextMessage = try signalEncrypt(
            message: messageBytes,
            for: remoteUserProtocolAddress,
            sessionStore: sessionStore,
            identityStore: identityStore,
            context: context
        )
        // Serialize and encode the cipher text message
        let serializedMessage = cipherTextMessage.serialize()
        return Data(serializedMessage).base64EncodedString()
    }

    public func decrypt(message: String) throws -> String? {
        guard let messageData = Data(base64Encoded: message) else {
            throw SignalError.invalidMessage("Invalid base64 encoded message")
        }

        let messageBytes = [UInt8](messageData)
        var decryptedBytes: [UInt8]

        do {
            // Try to parse as a normal SignalMessage first (most common scenario)
            let signalMessage = try SignalMessage(bytes: [UInt8](messageData))
            decryptedBytes = [UInt8](try signalDecrypt(
                message: signalMessage,
                from: remoteUserProtocolAddress,
                sessionStore: sessionStore,
                identityStore: identityStore,
                context: context
            ))
        } catch {
            // If parsing or decrypting as SignalMessage fails, fall back to PreKeySignalMessage
            do {
                let preKeySignalMessage = try PreKeySignalMessage(bytes: messageBytes)
                decryptedBytes = [UInt8](try signalDecryptPreKey(
                    message: preKeySignalMessage,
                    from: remoteUserProtocolAddress,
                    sessionStore: sessionStore,
                    identityStore: identityStore,
                    preKeyStore: preKeyStore,
                    signedPreKeyStore: signedPreKeyStore,
                    kyberPreKeyStore: kyberPreKeyStore,
                    context: context
                ))
            } catch let preKeyError {
                // PreKey may have been consumed by a previous message in this batch.
                // The first PreKeySignalMessage establishes the session via signalDecryptPreKey,
                // but subsequent messages from the same sender (before they receive our reply)
                // still arrive as PreKeySignalMessages. Since the session is already established,
                // we can extract the inner SignalMessage and decrypt it with the existing session.
                do {
                    let reParsed = try PreKeySignalMessage(bytes: messageBytes)
                    let innerSignalMessage = reParsed.signalMessage
                    decryptedBytes = [UInt8](try signalDecrypt(
                        message: innerSignalMessage,
                        from: remoteUserProtocolAddress,
                        sessionStore: sessionStore,
                        identityStore: identityStore,
                        context: context
                    ))
                } catch {
                    throw preKeyError
                }
            }
        }

        return String(bytes: decryptedBytes, encoding: .utf8)
    }

    /// Process remote user's preKey bundle to establish session
    private func processRemoteBundle(_ remoteUser: RemoteUser) throws {
        let preKeyBundle = try PreKeyBundle(
            registrationId: remoteUser.registrationId,
            deviceId: remoteUser.protocolAddress.deviceId,
            prekeyId: remoteUser.preKeyId,
            prekey: remoteUser.preKeyPublicKey,
            signedPrekeyId: remoteUser.signedPreKeyId,
            signedPrekey: remoteUser.signedPreKeyPublicKey,
            signedPrekeySignature: remoteUser.signedPreKeySignature,
            identity: remoteUser.identityKeyPairPublicKey,
            kyberPrekeyId: remoteUser.kyberPreKeyId,
            kyberPrekey: remoteUser.kyberPreKeyPublicKey,
            kyberPrekeySignature: remoteUser.kyberPreKeySignature
        )

        try processPreKeyBundle(
            preKeyBundle,
            for: remoteUser.protocolAddress,
            sessionStore: sessionStore,
            identityStore: identityStore,
            context: context
        )
    }
}